docs_src/design/adr/0019-Service-List.md

```diff
@@ -0,0 +1,87 @@
+# EdgeX Service List
+
+## Status
+**Draft as of April 25, 2021**
+Proposed for Jakarta or later release.
+
+## Context
+
+Several EdgeX Foundry services today require a list of the services that are part of an EdgeX deployment.  That is to say, that several EdgeX services need to know the names and locations of other EdgeX services that are or will be part of the EdgeX.
+
+The current list of needs include:
+
+- Security bootstrapping service requires a list of services in order to know which services need a file token for Vault access and to add shared secrets to Vault.
+- System management agent service requires a list of services in order to know which services it can provide status, configuration and metrics for client requestors.  The SMA also needs to know which services to include in any start/stop/restart operations.
+- The user interface needs to know which services, specifically device and application services, are available in order to provide interaction with those services, and potentially to display some metrics on operational services.
+- The CLI needs to know which services are operational (for some interaction) and which device services and application services to include in start/stop operations.
+
+In the future, there may be other internal or 3rd party needs to what services comprise an EdgeX deployment.
+
+### Current Situation
+
+Currently, the list of services is haphazardly managed by the needing service.
+
+- The security bootstrapping services are being built to use an environment variable to provide it with a list of services at startup.
+- The user interface and CLI do not deal with service lists today and assume all services are operational (with known service host and port defined in configuration).
+- The SMA has a static/hardcoded list built into the codebase.  Clearly the least desirable means of dealing with the list of services as this not only requires a restart of the service, but also code changes and redeployment.
+
+### Requirements
+
+1. It is desired that the service list be addressed in a consistent manner.  It is further desired that there be one single, authoritative source of the service list information for all EdgeX services and 3rd party requestors.  While the source of the service list may change under certain runtime circumstances (for example – whether operating with or without security), once EdgeX is running the authoritative source remains the same for all services.
+
+2. Any changes to the list of services must adhere to ACID transactional properties.  That is, any change to the list of services must be atomic, consistent, isolated and durable with regard to any operations on the service list.  In other words, no two services should ever be able to get a different list of services when making a request of EdgeX for the list of services.
```

Comment:
These two statements aren't the same.  "no two services should ever be able to get a different list of services when making a request of EdgeX for the list of services." means that the list is static. 

ACID-compliance means:
(A) Suppose there is an interface to add or delete a list of services.  Atomicity guarantees that no client sees a partially-executed list of adds or removes.
(C) Suppose there was a rule that removing a service from the list also removed it from the reverse proxy and that there is a 1:1 mapping between the two.  Consistency says that no client will ever observe a proxy rule without a accompanying service or a service without an accompanying proxy rule.  Both must be observed at the same time.
(I) If multiple concurrent updates to the service list are possible, a client will observe them as if they were applied in some sequential order.
(D) Means if the service list is modified, if the component in charge of the service list crashes and is restarted, the service list should be restored to its last modified state.

ACID properties are meaningless if the list of services is static.

Reply:
Bryon - what I meant by this (and have changed the language of this section to hopefully clarify) is that like in database transactions (that adhere to ACID), if service A makes a change to the the service list and then service B requests the service list, that service B should see the change made by service A with its request.  The service list is not static (or at least at this point in discussion we are thinking it will not be static).